src/NHibernate/Async/Hql/Ast/ANTLR/Exec/AbstractStatementExecutor.cs

```diff
@@ -68,9 +68,9 @@ protected virtual async Task DropTemporaryTableIfNecessaryAsync(IQueryable persi
 			{
 				IIsolatedWork work = new TmpIdTableDropIsolatedWork(persister, log, session);
 
-				if (ShouldIsolateTemporaryTableDDL())
+				if (ShouldIsolateTemporaryTableDDL() && session.ConnectionManager.CurrentTransaction != null)
 				{
-					session.ConnectionManager.Transaction.RegisterSynchronization(
+					session.ConnectionManager.CurrentTransaction.RegisterSynchronization(
```

Comment:
If you think that's it's worth adding `CurrentTransaction` to interface than we should use it instead  everywhere to make sure implementation is not ovveriden.

Reply:
Previous `Transaction` property implementation was not either (not `virtual`). So only a custom `ISession` may benefit of this change. (And of course only after we have put `CurrentTransaction` into the interface.)

Reply:
I am not committing your suggestion as it needs more changes to be effective. I will add another commit.

Reply:
In fact this change is invalid.
`session` is not an `ISession`, but an `ISessionImplementor`. It has never had a transaction property, and so the old code was just right going through the connection manager.

Reply:
> So only a custom ISession may benefit of this change. (And of course only after we have put CurrentTransaction into the interface.)

I don't get this point. If we add this to interface we can expect custom implementations. No?
So does it make sense to add `GetCurrentTransaction` to interface if it may results in different behavior - and here custom logic won't be used. Maybe it's enough to have it as extension methods?

> I am not committing your suggestion as it needs

I'm using suggested change just for illustration. I don''t expect it to be committed.

Reply:
We cannot do what you ask anyway. It was not supported by the previous `Transaction` property because it does not exist on the type this `session` variable is. And the change you are asking here is unrelated to what this PR aims to fix.

Reply:
My main point here -  it seems `ISession/IStatelessSession` is a wrong place for `Transaction/CurrentTransaction` property. It seems we should keep it simply as extension, so if you agree just remove TODO about adding new method to interface. And if someone needs extensibility - `CurrentTransaction` should be added directly to `ISessionImplementor` (with different PR)

P.S. This PR still contains breaking changes in `IStatelessSession`.

Reply:
Fixed, previous commits squashed, new changes are in the fixup commit.
Some glitches in previous fixup commits have been adjusted while squashing.